code/modules/hydroponics/compost_bin.dm

```diff
@@ -0,0 +1,192 @@
+				break
+
+			//create biomass from plant nutriment and plant matter
+			var/plant_biomass = G.reagents.get_reagent_amount("nutriment") + G.reagents.get_reagent_amount("plantmatter")
+
+			biomass += min(max(plant_biomass, 0.1) * 10,biomass_capacity-biomass)
+
+			PB.remove_from_storage(G, src)
+
+		// start composting after plants are inserted
+		compost()
+
+		if(biomass < biomass_capacity)
+			to_chat(user, "<span class='info'>You empty [PB] into [src].</span>")
+		else
+			to_chat(user, "<span class='info'>You fill [src] to its capacity.</span>")
+
+		SStgui.update_uis(src)
+		return TRUE
+
+	else if(istype(O, /obj/item/reagent_containers/food/snacks/grown))
+		if(biomass >= biomass_capacity)
+			to_chat(user, "<span class='warning'>[src] can't hold any more plants!</span>")
+			return
+		if(!user.unEquip(O))
+			return
+
+		O.forceMove(src)
+		//make biomass from nutriment and plant matter
+		var/plant_biomass = O.reagents.get_reagent_amount("nutriment") + O.reagents.get_reagent_amount("plantmatter")
+		biomass += min(max(plant_biomass, 0.1) * 10,biomass_capacity-biomass)
+		qdel(O)
+		// start composting after plants are inserted
+		compost()
+		to_chat(user, "<span class='info'>You put [O] in [src].</span>")
+		SStgui.update_uis(src)
+		return TRUE
+	else
+		to_chat(user, "<span class='warning'>You cannot put this in [name]!</span>")
+	SStgui.update_uis(src)
+	update_icon_state()
+
+
+/obj/machinery/compost_bin/ui_data(mob/user)
+	var/list/data = list(
+		"biomass" = biomass,
+		"biomass_capacity" = biomass_capacity,
+		"compost" = compost,
+		"compost_capacity" = compost_capacity,
+		)
+	return data
+
+/obj/machinery/compost_bin/attack_hand(mob/user)
+	ui_interact(user)
+
+
+/obj/machinery/compost_bin/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "CompostBin", "CompostBin", 390, 300, master_ui, state)
+		ui.set_autoupdate(FALSE)
+		ui.open()
+
+
+// Start composting if there is enough biomass and space for compost
+/obj/machinery/compost_bin/proc/compost()
+	// Prevents the compost bin from starting to compost again while already composting
+	if(composting == TRUE)
+		return
+	composting = TRUE
+	if(compost >= compost_capacity || biomass <= 0)
+		composting = FALSE
+		return
+	else
+		addtimer(CALLBACK(src, PROC_REF(convert_biomass)), 100)
+
+// Convert biomass to compost, then continue composting
+/obj/machinery/compost_bin/proc/convert_biomass()
+	//converts 20% of the biomass to compost each cycle, unless there isn't enough comopst space or there is 10 or under biomass
+	var/conversion_amount = max(min(DECAY*biomass,compost_capacity - compost),min(MIN_CONVERSION,biomass))
+	biomass -= conversion_amount
+	compost += conversion_amount
+	update_icon_state()
+	SStgui.update_uis(src)
+	composting = FALSE
+	compost()
```

Comment:
You could check here if there's enough to compost, and if so, already make a callback to this proc

Reply:
I didn't want to duplicate code there, but I'll give it a shot.